src/Microsoft.SqlTools.ServiceLayer/Connection/ConnectionService.cs

```diff
@@ -188,44 +188,78 @@ public ConnectionCompleteParams ValidateConnectParams(ConnectParams connectionPa
             // If there is no ConnectionInfo in the map, create a new ConnectionInfo, 
             // but wait until later when we are connected to add it to the map.
             ConnectionInfo connectionInfo;
+            bool connectionChanged = false;
             if (!ownerToConnectionMap.TryGetValue(connectionParams.OwnerUri, out connectionInfo))
             {
                 connectionInfo = new ConnectionInfo(ConnectionFactory, connectionParams.OwnerUri, connectionParams.Connection);
             }
-
-            // Resolve if it is an existing connection
-            // Disconnect active connection if the URI is already connected for this connection type
-            DbConnection existingConnection;
-            if (connectionInfo.TryGetConnection(connectionParams.Type, out existingConnection))
-            {
-                var disconnectParams = new DisconnectParams()
-                {
-                    OwnerUri = connectionParams.OwnerUri, 
-                    Type = connectionParams.Type
-                };
-                Disconnect(disconnectParams);
+            else if (IsConnectionChanged(connectionParams, connectionInfo))
+            {
+                // We are actively changing the connection information for this connection. We must disconnect
+                // all active connections, since it represents a full context change
+                connectionChanged = true;
             }
 
-            // Try to open a connection with the given ConnectParams
+            DisconnectExistingConnectionIfNeeded(connectionParams, connectionInfo, disconnectAll: connectionChanged);
+
+            if (connectionChanged)
+            {
+                connectionInfo = new ConnectionInfo(ConnectionFactory, connectionParams.OwnerUri, connectionParams.Connection);
+            }
+
+            // Try to open a connection with the given ConnectParams
             ConnectionCompleteParams response = await TryOpenConnection(connectionInfo, connectionParams);
             if (response != null)
             {
                 return response;
-            }
-
-            // If this is the first connection for this URI, add the ConnectionInfo to the map
-            if (!ownerToConnectionMap.ContainsKey(connectionParams.OwnerUri))
+            }
+
+            // If this is the first connection for this URI, add the ConnectionInfo to the map
+            bool addToMap = connectionChanged || !ownerToConnectionMap.ContainsKey(connectionParams.OwnerUri);
+            if (addToMap)
             {
                 ownerToConnectionMap[connectionParams.OwnerUri] = connectionInfo;
```

Comment:
This, in particular, isn't safe since the contains key and add to map actions aren't performed in an atomic block. I'd suggest using a try add, since that will not add if the existing item already exists.